File: src/windows/Maple.WinUI/MapleUI/Molecules/MuiToolbar.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>
    /// Maple.UI Toolbar molecule (unified-component-catalog.md §2.5,
    /// "Toolbar" row: "Row of actions with overflow", built from Action
    /// Button, Divider, Icon) — a row of <see cref="MuiActionButton"/>s;
    /// once the item entries exceed <see cref="MaxVisible"/> the rest
    /// collapse behind a trailing overflow button that opens an
    /// <see cref="MuiPopover"/> list, via <see cref="MuiToolbarMath"/>.
    ///
    /// Ports `mui-toolbar.component.ts`'s split contract. Toolbar entries
    /// carry no persisted "active" state (unlike a tool dock's own use of
    /// Action Button for the currently-armed tool) — each press is
    /// momentary, so the button's <see cref="MuiActionButton.Selected"/>
    /// toggle (inherited from wrapping a <see cref="Microsoft.UI.Xaml.Controls.Primitives.ToggleButton"/>)
    /// is reset immediately after firing <see cref="ItemSelected"/>.
    /// </summary>
    public sealed class MuiToolbar : ContentControl
    {
        public static readonly DependencyProperty EntriesProperty =
            DependencyProperty.Register(nameof(Entries), typeof(IReadOnlyList<MuiToolbarEntry>), typeof(MuiToolbar),
                new PropertyMetadata(null, (d, _) => ((MuiToolbar)d).Rebuild()));

        public static readonly DependencyProperty MaxVisibleProperty =
            DependencyProperty.Register(nameof(MaxVisible), typeof(int), typeof(MuiToolbar),
                new PropertyMetadata(int.MaxValue, (d, _) => ((MuiToolbar)d).Rebuild()));

        public IReadOnlyList<MuiToolbarEntry>? Entries
        {
            get => (IReadOnlyList<MuiToolbarEntry>?)GetValue(EntriesProperty);
            set => SetValue(EntriesProperty, value);
        }

        /// <summary>Item entries beyond this count move into the overflow
        /// popover. Dividers don't count against the budget.</summary>
        public int MaxVisible
        {
            get => (int)GetValue(MaxVisibleProperty);
            set => SetValue(MaxVisibleProperty, value);
        }

        public event EventHandler<string>? ItemSelected;

        private readonly StackPanel _root = new() { Orientation = Orientation.Horizontal, Spacing = 4 };
        private readonly StackPanel _visibleRow = new() { Orientation = Orientation.Horizontal, Spacing = 4 };
        private readonly Grid _overflowHost = new();
        private readonly MuiButton _overflowTrigger = new() { Variant = MuiButtonVariant.Ghost, ButtonSize = MuiButtonSize.Sm, IconName = "ellipsis-horizontal" };
        private readonly MuiPopover _overflowPopover = new();
        private readonly StackPanel _overflowList = new() { Orientation = Orientation.Vertical, Spacing = 2, MinWidth = 180 };

        public MuiToolbar()
        {
            _overflowHost.Children.Add(_overflowTrigger);
            _overflowHost.Children.Add(_overflowPopover);

            _root.Children.Add(_visibleRow);
            _root.Children.Add(_overflowHost);
            Content = _root;
            IsTabStop = false;

            _overflowPopover.PanelContent = _overflowList;
            _overflowTrigger.Click += (_, _) => _overflowPopover.IsOpen = !_overflowPopover.IsOpen;
            _overflowPopover.CloseRequested += (_, _) => _overflowPopover.IsOpen = false;

            Rebuild();
        }

        private static Brush R(string key) => (Brush)Application.Current.Resources[key];

        private void Press(MuiToolbarItem item)
        {
            if (item.Disabled) return;
            ItemSelected?.Invoke(this, item.Id);
        }

        private void SelectOverflow(MuiToolbarItem item)
        {
            if (item.Disabled) return;
            ItemSelected?.Invoke(this, item.Id);
            _overflowPopover.IsOpen = false;
        }

        private void Rebuild()
        {
            var split = MuiToolbarMath.Split(Entries ?? Array.Empty<MuiToolbarEntry>(), MaxVisible);

            _visibleRow.Children.Clear();
            foreach (var entry in split.Visible)
            {
                if (entry.IsDivider)
                {
                    _visibleRow.Children.Add(new MuiDivider { Orientation = MuiDividerOrientation.Vertical, Margin = new Thickness(2, 0, 2, 0) });
                    continue;
                }

                var item = entry.Item!;
                var button = new MuiActionButton { IconName = item.IconName, Label = item.Label, IsEnabled = !item.Disabled };
                button.Click += (_, _) =>
                {
                    button.Selected = false;
                    Press(item);
                };
                _visibleRow.Children.Add(button);
            }

            _overflowTrigger.Visibility = split.Overflow.Count > 0 ? Visibility.Visible : Visibility.Collapsed;

            _overflowList.Children.Clear();
            foreach (var item in split.Overflow)
                _overflowList.Children.Add(BuildOverflowRow(item));
        }

        private Border BuildOverflowRow(MuiToolbarItem item)
        {
            var content = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8 };
            content.Children.Add(new MuiIcon { IconName = item.IconName, Size = MuiIconSize.Sm16 });
            content.Children.Add(new MuiText { Text = item.Label, Variant = MuiTextVariant.RowLabel });

            var row = new Border
            {
                CornerRadius = new CornerRadius(6),
                Padding = new Thickness(10, 7, 10, 7),
                Child = content,
                Opacity = item.Disabled ? 0.45 : 1.0,
            };
            row.PointerPressed += (_, _) => SelectOverflow(item);
            row.PointerEntered += (_, _) => row.Background = R("MapleSurfaceHover");
            row.PointerExited += (_, _) => row.Background = new SolidColorBrush(Microsoft.UI.Colors.Transparent);
            return row;
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/Molecules/MuiToolbar.swift
// MuiToolbar.swift — Maple UI Molecules-L1 (unified-component-catalog.md
// §2.5; Built from: Action Button, Divider, Icon). A row of action
// buttons; once the item entries exceed `maxVisible` the rest collapse
// behind a trailing overflow button that opens a popover list.

import SwiftUI

public struct MuiToolbarActionItem: Identifiable, Sendable {
    public let id: String
    public let icon: String
    public let label: String
    public let disabled: Bool

    public init(id: String, icon: String, label: String, disabled: Bool = false) {
        self.id = id
        self.icon = icon
        self.label = label
        self.disabled = disabled
    }
}

public enum MuiToolbarEntry: Sendable {
    case item(MuiToolbarActionItem)
    case divider
}

public struct MuiToolbar: View {
    public let entries: [MuiToolbarEntry]
    /// Item entries beyond this count move into the overflow popover.
    /// Dividers don't count against the budget.
    public let maxVisible: Int
    public let itemSelected: (String) -> Void

    @State private var overflowOpen = false

    public init(entries: [MuiToolbarEntry], maxVisible: Int = .max, itemSelected: @escaping (String) -> Void) {
        self.entries = entries
        self.maxVisible = maxVisible
        self.itemSelected = itemSelected
    }

    /// Splits `entries` into the visible row and the overflow list — a
    /// divider only earns its place while still filling the visible row;
    /// once overflow has started, trailing dividers add nothing to either
    /// list. Public + static so this is unit-testable without rendering a
    /// view.
    public static func split(
        entries: [MuiToolbarEntry],
        maxVisible: Int
    ) -> (visible: [MuiToolbarEntry], overflow: [MuiToolbarActionItem]) {
        var visible: [MuiToolbarEntry] = []
        var overflow: [MuiToolbarActionItem] = []
        var itemCount = 0
        for entry in entries {
            switch entry {
            case .item(let item):
                if itemCount < maxVisible {
                    visible.append(entry)
                    itemCount += 1
                } else {
                    overflow.append(item)
                }
            case .divider:
                if overflow.isEmpty { visible.append(entry) }
            }
        }
        return (visible, overflow)
    }

    public var body: some View {
        let split = Self.split(entries: entries, maxVisible: maxVisible)

        HStack(spacing: MuiTokens.spacingXs) {
            ForEach(Array(split.visible.enumerated()), id: \.offset) { _, entry in
                switch entry {
                case .item(let item):
                    MuiActionButton(icon: item.icon, label: item.label, size: .sm, disabled: item.disabled) {
                        press(item)
                    }
                case .divider:
                    MuiDivider(orientation: .vertical).frame(height: 20)
                }
            }

            if !split.overflow.isEmpty {
                MuiActionButton(icon: "ellipsis", label: "More", size: .sm) {
                    overflowOpen.toggle()
                }
                .muiPopover(isPresented: overflowOpen, placement: .bottom, closeRequested: { overflowOpen = false }) {
                    VStack(alignment: .leading, spacing: 2) {
                        ForEach(split.overflow) { item in
                            Button {
                                selectOverflow(item)
                            } label: {
                                HStack(spacing: MuiTokens.spacingSm) {
                                    MuiIcon(name: item.icon, size: .sm, color: MuiTokens.textMuted)
                                    MuiText(item.label, variant: .rowLabel)
                                    Spacer(minLength: MuiTokens.spacingSm)
                                }
                                .padding(.horizontal, MuiTokens.spacingSm)
                                .padding(.vertical, MuiTokens.spacingXs)
                                .frame(minHeight: 32)
                                .frame(maxWidth: .infinity, alignment: .leading)
                            }
                            .buttonStyle(.plain)
                            .disabled(item.disabled)
                            .opacity(item.disabled ? 0.45 : 1)
                        }
                    }
                    .frame(minWidth: 160)
                }
            }
        }
    }

    private func press(_ item: MuiToolbarActionItem) {
        guard !item.disabled else { return }
        itemSelected(item.id)
    }

    private func selectOverflow(_ item: MuiToolbarActionItem) {
        guard !item.disabled else { return }
        itemSelected(item.id)
        overflowOpen = false
    }
}

#Preview("MuiToolbar") {
    VStack(spacing: 16) {
        MuiToolbar(
            entries: [
                .item(MuiToolbarActionItem(id: "crop", icon: "crop", label: "Crop")),
                .item(MuiToolbarActionItem(id: "rotate", icon: "rotate.right", label: "Rotate")),
                .divider,
                .item(MuiToolbarActionItem(id: "export", icon: "square.and.arrow.up", label: "Export")),
            ],
            itemSelected: { _ in }
        )

        MuiToolbar(
            entries: [
                .item(MuiToolbarActionItem(id: "crop", icon: "crop", label: "Crop")),
                .item(MuiToolbarActionItem(id: "rotate", icon: "rotate.right", label: "Rotate")),
                .item(MuiToolbarActionItem(id: "export", icon: "square.and.arrow.up", label: "Export")),
                .item(MuiToolbarActionItem(id: "share", icon: "person.2", label: "Share")),
            ],
            maxVisible: 2,
            itemSelected: { _ in }
        )
    }
    .padding(80)
    .background(MuiTokens.bg)
}
